File: src/Sorting/BubbleSort.php
<?php

namespace GulDmitry\Algorithms\Sorting;

/**
 * O(n^2)
 * Worst case - (N-1)*(N/2)
 */
class BubbleSort
{
    public function sort(array $raw, bool $ascDirection = true): array
    {
        $iterator = new \ArrayIterator($raw);
        do {
            $isSwapped = false;

            while ($iterator->valid()) {
                $currentKey = $iterator->key();
                $currentValue = $iterator->current();

                $iterator->next();
                if (!$iterator->valid()) {
                    continue;
                }
                $nextKey = $iterator->key();
                $nextValue = $iterator->current();

                $compareExpression = $ascDirection ? $currentValue > $nextValue : $currentValue < $nextValue;
                if ($compareExpression) {
                    list($iterator[$currentKey], $iterator[$nextKey]) = [$iterator[$nextKey], $iterator[$currentKey]];
                    $isSwapped = true;
                }
            }
            $iterator->rewind();
        } while ($isSwapped);

        return (array)$iterator;
    }
}

File: src/Sorting/bubble_sort.go
package Sorting

func Sort(raw []int, ascDirection bool) []int {
	length := len(raw)

	rawCopy := make([]int, length)
	copy(rawCopy, raw)
	isSwapped := false
	var compareExpression bool

	for ok := true; ok; ok = isSwapped {
		isSwapped = false
		for k, v := range rawCopy {

			if length == k+1 {
				continue
			}

			nextV := rawCopy[k+1]

			if ascDirection {
				compareExpression = v > nextV
			} else {
				compareExpression = v < nextV
			}
			if compareExpression {
				rawCopy[k], rawCopy[k+1] = rawCopy[k+1], rawCopy[k]
				isSwapped = true
			}
		}
	}
	return rawCopy
}
